### PR TITLE
svm: integrate Firedancer's fuzz harness for instructions

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
We need a reliable way to test for regressions in many SVM components as we continue the work to enable direct mapping, ABI v2, and more. We removed an antiquated setup for this in #7322, but we need a proper replacement.

The Firedancer team has been maintaining such a tool standalone, rebasing and fixing breakage with every Agave release. We can just bring that fuzzing harness in-house.

#### Summary of Changes
Extracts the instruction (`instr`) fuzz harness from Firedancer's [SolFuzz-Agave](https://github.com/firedancer-io/solfuzz-agave) and integrates it alongside SVM.

This harness will be be used to test the SVM for stability and catch any edge cases with new features. It will also be migrated to the new SVM repo when the time comes.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "svm-fuzz-harness/fixture/protosol"]
+	path = svm-fuzz-harness/fixture/protosol
+	url = https://github.com/firedancer-io/protosol.git
```
Comment (buffalojoec) [inline .gitmodules]: @mjain-jump my idea here was to use a git submodule, but maybe it's better if we publish and version the protobufs somewhere?

File: svm-fuzz-harness/scripts/run_test_vectors.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+mkdir -p dump
+
+
+if [ ! -d dump/test-vectors ]; then
+  cd dump
+  git clone --depth=1 -q https://github.com/firedancer-io/test-vectors.git
+  cd ..
+else
```
Comment (buffalojoec) [inline svm-fuzz-harness/scripts/run_test_vectors.sh:24]: @mjain-jump it would be great if these were compressed into a zip and we could just download that. We probably also want to version the zips according to the Agave version.

File: svm-fuzz-harness/scripts/run_test_vectors.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  git pull -q
+  cd ../..
+fi
+
+
+# Skip only the programs that have been migrated to Core BPF
+find dump/test-vectors/instr/fixtures -type f -name '*.fix' | \
+  grep -v '/stake/' | \
+  grep -v '/config/' | \
+  grep -v '/address-lookup-table/' | \
```
Comment (buffalojoec) [inline svm-fuzz-harness/scripts/run_test_vectors.sh]: @mjain-jump I've been omitting these folders since these programs are no longer builtins. Should we just remove them from the vectors?

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 2 above)
Comment (Rexicon226) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: I would love this as well. We test the entire test-vectors suite in [Sig's](https://github.com/Syndica/sig) CI, which requires us to pull down the repo. We don't want to track master because that will cause random CI failures, and afaik git doesn't have a way to pull down a specific commit that isn't the HEAD (--depth 1). So for now we just use --depth 1 + checkout. It would be great if we could have rolling releases of the zips that contain test vectors generated for a specific Agave version. Rolling in the sense that when new fixtures are added to the test-vectors, those zips are regenerated in a staged manner, cuz another implementation might also fail the new fixture.

Comment (mjain-jump) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: The plan is to start cutting releases for test vectors every Agave minor release ✅

File: svm-fuzz-harness/scripts/run_test_vectors.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  xargs -P 32 -I {} ../target/release/test_exec_instr {} > "$LOG_PATH/test_exec_instr.log" 2>&1
```
Comment (mjain-jump) [inline svm-fuzz-harness/scripts/run_test_vectors.sh]: add `-n 1000` - we had some CI slowdowns due to uncapped argument counts before

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 3 above)
Comment (mjain-jump) [reply svm-fuzz-harness/scripts/run_test_vectors.sh]: just pushed a migration for these programs to be BPF. you should be able to get rid of these now!

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 2 above)
Comment (LucasSte) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: > The plan is to start cutting releases for test vectors every Agave minor release ✅

If that is the case, then the test vectors cannot run on master, right? Or am I mistaken?

Comment (buffalojoec) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: I think this is a good point. We could run them in consensus mode (ignore error codes and maybe CUs for instructions, for example). However, we might be able to tweak it a bit more.

The way this harness should probably work is that features should be disabled by default unless specified in the input message. When we add new codepaths, we shouldn't ever trip existing vectors, and we should be able to generate new vectors from a corpus of tests.

Comment (mjain-jump) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: Few options I had in mind:
1. We tie test-vectors to Agave versions (i.e. different test-vectors releases for 3.1.0, 3.1.1, etc) and run them on master
2. We tie test-vectors to a commit SHA on Agave master (as a checkpoint) and regenerate them every X days / weeks, or as needed (i.e. feature gets rekeyed). This might be more desirable and catch more breaking changes that may get accidentally pushed.

Open to hearing other ideas as well. I'm a bit opposed to the consensus-mode idea just because a lot of pain has come from even refactoring error codes that change in the runtime.

File: svm-fuzz-harness/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+version = { workspace = true }
+authors = { workspace = true }
+repository = { workspace = true }
+homepage = { workspace = true }
+license = { workspace = true }
+edition = { workspace = true }
+publish = false
+
+[lib]
+crate-type = ["cdylib", "rlib"]
```
Comment (LucasSte) [inline svm-fuzz-harness/Cargo.toml:14]: Do we need it to be an rlib?

File: svm-fuzz-harness/scripts/run_test_vectors.sh [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+
+
+
+mkdir -p dump
+
+
+if [ ! -d dump/test-vectors ]; then
+  cd dump
+  git clone --depth=1 -q https://github.com/firedancer-io/test-vectors.git
```
Comment (LucasSte) [inline svm-fuzz-harness/scripts/run_test_vectors.sh:22]: You could potentially specify the branch to clone to make the command faster.

File: svm-fuzz-harness/scripts/run_test_vectors.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  cd ..
+else
+  cd dump/test-vectors
+  git pull -q
```
Comment (LucasSte) [inline svm-fuzz-harness/scripts/run_test_vectors.sh:26]: Here as well.

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 2 above)
Comment (LucasSte) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: I believe we should run the tests on master to prevent regressions. If something changes behavior, it should be behind a feature gate, right?

The problem I see right now would be when we are cleaning up feature gates. The PR would be blocked until we can update the vectors. Perhaps, we could automate that as well?

Comment (buffalojoec) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: > I believe we should run the tests on master to prevent regressions. If something changes behavior, it should be behind a feature gate, right?

Yes, if it affects consensus. Instruction and transaction error codes should only change from a feature gate, but lower-level stuff like syscalls and VM instructions can change codes if they don't bubble up to the block.

Comment (LucasSte) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: What affects consensus is only the difference between success and error. We actually allow error codes to change without a feature gate. That might create more burden when running vectors on master.

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 6 above)
Comment (buffalojoec) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:22]: @mjain-jump sounds like we need a better story for the vectors before this goes in. I know we have some discussion going in another thread about this.

I'm still in favor of doing releases on the `test-vectors` repository with `.zip` folders (or something more compact?). We could do a versioning scheme like `agave-3.0.0-a77ea93` (version + commit SHA).

File: svm-fuzz-harness/Cargo.toml (hunk 5 above)
Comment (buffalojoec) [reply svm-fuzz-harness/Cargo.toml:14]: Yes the `bin` target imports the library for use in Rust.

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 6 above)
Comment (LucasSte) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:22]: There could be a job to publish the vector that runs on every commit made into the test-vectors master branch. Alternatively, it could be manually triggered.

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 2 above)
Comment (mjain-jump) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:24]: > The PR would be blocked until we can update the vectors. Perhaps, we could automate that as well?

We have tools to do this in under 2 mins. We can coordinate this very easily.

> What affects consensus is only the difference between success and error

Agreed, but we have had a lot of pain with matching error codes that change as well. We would like to compare down to error codes too if that's possible.

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 6 above)
Comment (mjain-jump) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:22]: In FD we use a commit SHA for the test-vectors commit to use, and it checks that commit SHA out. Every time the test-vectors are updated we update the SHA.

> I'm still in favor of doing releases on the test-vectors repository with .zip folders (or something more compact?). We could do a versioning scheme like agave-3.0.0-a77ea93 (version + commit SHA).

I'm in favor of this. We can do thing like you mentioned, where its `agave-<version>-<test-vectors-commit-sha>` which should be easy to track and update.

**Review by mjain-jump — APPROVED** [review]

**Review by LucasSte — APPROVED** [review]

**Review by Lichtso — APPROVED** [review]

File: svm-fuzz-harness/scripts/run_test_vectors.sh (hunk 6 above)
Comment (steviez) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:22]: From an old [Discord discussion](https://discord.com/channels/428295358100013066/560503042458517505/1412991525398249513), we backed stuff out with https://github.com/anza-xyz/agave/pull/7322 due to issues with cloning the repo.

I see there is some discussion here but was a decision reached / are we confident that we will not start hitting these issues again ?

Comment (buffalojoec) [reply svm-fuzz-harness/scripts/run_test_vectors.sh:22]: We actually haven't hooked it up to CI yet. The general consensus was to zip vectors up and just download the zipped folder. After 3.1 we'll hook up CI runs.
